proto/backend.proto

```diff
@@ -0,0 +1,212 @@
+syntax = "proto3";
+import "google/protobuf/empty.proto";
+package grafana.plugin;
+
+//-----------------------------------------------
+// Common
+//-----------------------------------------------
+
+message PluginConfig {
+  int64 id = 1;
+  int64 orgId = 2;
+  string name = 3;
+  string type = 4;
+  string url = 5;
+  string jsonData = 6;
+  map<string,string> decryptedSecureJsonData = 7;
+}
+
+
+//-----------------------------------------------
+// All Backend Plugins Implement
+//-----------------------------------------------
+
+service BackendPlugin {
+  // Get the plugin status
+  rpc Check(PluginStatusRequest) returns (PluginStatusResponse);
+
+  // HTTP Style request
+  rpc Resource(ResourceRequest) returns (ResourceResponse);
+
+  // Well typed query interface
+  rpc DataQuery(DataQueryRequest) returns (DataQueryResponse);
+
+}
+
+//-----------------------------------------------
+// Plugin Status Check
+//-----------------------------------------------
+
+message PluginStatusRequest {
+  // Plugin Configuration
+  PluginConfig config = 1;
+
+  // Environment info
+  map<string,string> headers = 2;
+
+  enum FetchInfo {
+    STATUS = 0;  // nothing
+    API = 1;     // OpenAPI description
+    METRICS = 2; // Prometheus style metrics
+    DEBUG = 3;   // JSON Debug Info (admin+ view)
+  }
+
+  // Get an OpenAPI description (viewer+)
+  FetchInfo fetch = 3;
+}
+
+message PluginStatusResponse {
+  enum PluginStatus {
+    UNKNOWN = 0; // starting etc
+    OK = 1;
+    ERROR = 2;
+  }
+
+  PluginStatus status = 1;
+  string info = 2;
+}
+
+//-----------------------------------------------
+// Data Query
+//-----------------------------------------------
+
+message TimeRange {
+  int64 fromEpochMS = 1;
+  int64 toEpochMS = 2;
+}
+
+message DataQuery {
+  string refId = 1;
+  int64 maxDataPoints = 2;
+  int64 intervalMS = 3;
+  TimeRange timeRange = 4; // NOTE: this moved from QueryRequest in v1
+  bytes json = 5; // Arbitrary structure (model?  modelJson in v1)
+}
+
+message DataQueryRequest {
+  // Plugin Configuration
+  PluginConfig config = 1;
+
+  // Environment info
+  map<string,string> headers = 2;
+
+  // List of queries
+  repeated DataQuery queries = 3;
+}
+
+message DataQueryResponse {
+  // Arrow encoded DataFrames
+  // Each frame encodes its own: Errors, meta, and refId
+  repeated bytes frames = 1;
+
+  // Additional response metadata
+  map<string,string> metadata = 2;
+}
+
+//-----------------------------------------------
+// HTTP style request
+//-----------------------------------------------
+
+message ResourceRequest {
+  // Plugin Configuration
+  PluginConfig config = 1;
+
+  
+  // HTTP Style parameters
+  map<string,string> headers = 2; // Request Headers
```

Comment:
@ryantxu We have a pattern of "headers" being for "Environment info" which I'm confused on.

 1. I'm not clear on what will go in those headers, and have a concern about Plugin authors conflicting in that space with development of the plugin platform/grafana itself.
 2. In this Case with a Resource (HTTP Style) request, it gets confusing because there are Headers you might want proxy through your plugin to source system (say a third part rest API) and whatever we are using "environment info" info for.

Reply:
Thinking of them as HTTP headers is fine.  The environment info we should pass though should look like HTTP headers (Authorization: XXXX etc)

We can (and should) also use the builtin gRPC meadata scheme for some environment stuff.  Like perhaps when the thing starts passing though license info etc?